sdk/core/core-http/src/util/exponentialBackOffRetry.ts

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { HttpOperationResponse } from "../coreHttp";
+
+export const DEFAULT_CLIENT_RETRY_COUNT = 3;
+// intervals are in ms
+export const DEFAULT_CLIENT_RETRY_INTERVAL = 1000 * 30;
+export const DEFAULT_CLIENT_MAX_RETRY_INTERVAL = 1000 * 90;
+export const DEFAULT_CLIENT_MIN_RETRY_INTERVAL = 1000 * 3;
+
+export function isNumber(n: any): n is number {
+  return typeof n === "number";
+}
+export interface RetryData {
+  retryCount: number;
+  retryInterval: number;
+  error?: RetryError;
+}
+
+export interface RetryError extends Error {
+  message: string;
+  code?: string;
+  innerError?: RetryError;
+}
+
+/**
+ * @internal
+ * Determines if the operation should be retried.
+ *
+ * @param {number} retryLimit Specifies the max number of retries.
+ * @param {(response?: HttpOperationResponse, error?: RetryError) => boolean} predicate Initial chekck on whether to retry based on given responses or errors
+ * @param {RetryData} retryData  The retry data.
+ * @return {boolean} True if the operation qualifies for a retry; false otherwise.
+ */
+export function shouldRetry(
+  retryLimit: number,
+  predicate: (response?: HttpOperationResponse, error?: RetryError) => boolean,
+  retryData: RetryData,
+  response?: HttpOperationResponse,
+  error?: RetryError
+): boolean {
+  if (!predicate(response, error)) {
+    return false;
+  }
+
+  let currentCount: number;
+  if (!retryData) {
```

Comment:
I think we can enforce in calling code that this is never falsy so this check isn't needed?

Reply:
Good point! We call `updateRetryData()` before this. With the default value this should be valid.